### ISSUE TITLE
Use sts Ansible module?

### ISSUE THREAD

**Issue by viper233** [docker-production-aws/aws-sts#2]
https://github.com/docker-production-aws/aws-sts/blob/2b9c65777d1a45b0b48af1fa8b308b57d190705c/tasks/main.yml#L40

http://docs.ansible.com/ansible/latest/sts_assume_role_module.html There is a module to do this.

e.g.
```
- name: Assume role to get credentials for IAM role setup.
  sts_assume_role:
    profile: "{{ assume_aws_role_aws_profile }}"
    region: "{{ assume_aws_role_aws_region }}"
    role_arn: "{{ assume_aws_role_role_arn_to_assume }}"
    role_session_name: "{{ assume_aws_role_role_session_name }}"
  register: assumed_role

- name: seeing what is set
  debug:
    var: assumed_role
    verbosity: 2

- name: Set the assumed role aws credentials as the ones to be used.
  set_fact:
    aws_access_key: "{{ assumed_role.sts_creds.access_key }}"
    aws_secret_key: "{{ assumed_role.sts_creds.secret_key }}"
    security_token: "{{ assumed_role.sts_creds.session_token }}"
```

**Comment by mixja** [comment]
Hi - thanks for the feedback.

A major limitation of the Ansible STS module is that it is using the older `boto` AWS SDK, which doesn't include support for MFA, hence the use of the plain old CLI command (which uses `boto3` under then hood).

Also, even if the STS module did support `boto3`, another problem is that the AWS credential cache is not a feature of `boto3` and instead is a feature of the AWS CLI.  What this means is that you would be prompted for MFA credentials every time you ran the playbook, where as by using the CLI we get the benefits of credential caching (meaning you only need to re-enter MFA credentials when your temporary session token expires).

So in summary, the choice to use the AWS CLI here was driven by a requirement to support MFA and running playbooks locally with an optimal experience.

**Comment by viper233** [comment]
Hi
Thanks for the comments and feedback. I've implemented it with MFA support and *you are right*, it doesn't cache the token and requires the MFA to be entered each time, which can be a pain.

Now the cache is only a [file](http://docs.aws.amazon.com/cli/latest/userguide/cli-roles.html#cli-roles-cache). I should be possible to supply this as a parameter to the module if it was re-written with boto3, which is in the works. [ansible/ansible/pull/325469]. No mention of a cache file http://botocore.readthedocs.io/en/latest/reference/services/sts.html
 
thanks again, enjoy your weekend

**Comment by viper233** [comment]
Support for a cache file just got merged boto/botocore/pull/1157. Going off to enjoy my weekend now

**Comment by mixja** [comment]
That's great news that session caching has made its way into `botocore` - historically I even [created a wrapper around boto3 to deal with this](https://github.com/mixja/boto3-session-cache) so having native support will be much better.

Once this feature is available via `boto3`, it will raise the question of whether or not you even need the  `aws-sts` Ansible role, as the Ansible CloudFormation module used in the aws-cloudformation role uses `boto3` from Ansible 2.4 onwards.  The `aws-sts` role/sts module may be useful on build systems where you want the build system to dynamically assume a role linked to the build job, but I see this as being the exception rather than the norm.
